### ISSUE TITLE
Ex2007sp3rp17 edgetransport unhandled exception on service stop

### ISSUE THREAD

**Issue by AlxVD** [Pro/dkim-exchange#104]
1. Using v.2.1.7 (for Ex2007sp3rp17 on WinSrv2008R2, .net4.0).
2. Trying to stop or restart MSExchangeTransport service.
3. edgetransport.exe dies with an unhandled exception:
![ue](https://cloud.githubusercontent.com/assets/15388258/10835278/c766a902-7eb1-11e5-80e3-14cade55019b.png)

And here is Watson zipped dump/report files: https://cloud.mail.ru/public/6wBS/RowUBE1Ve

**Comment by Pro** [comment]
Is there any Message in the Event Log from the DKIM Signer?

**Comment by AlxVD** [comment]
No. It goes like this in App log (even with LogLevel 4):
```
MSExchange TransportService	The service is trying to stop.
ESE							edgetransport (6908) Sender Reputation Database: The database engine stopped the instance (2). 
ESE							edgetransport (6908) IP Filtering Database: The database engine stopped the instance (1). 
ESE							edgetransport (6908) Transport Mail Database: The database engine stopped the instance (0). 
MSExchangeTransport			Queuing status at shutdown - Submission queue: 0 message(s); Delivery queues: 5 message(s); Unreachable queue: 0 message(s); Poison queue: 0 message(s).
Windows Error Reporting		Fault bucket , type 0
				Event Name: CLR20r3
				Response: Not available
				Cab Id: 0

				Problem signature:
				P1: edgetransport.exe
				P2: 8.3.406.0
				P3: 54ddb00e
				P4: mscorlib
				P5: 2.0.0.0
				P6: 53a11de1
				P7: 4e39
				P8: 6
				P9: UDTA330IDOBH2ROZ2AYVLCELAG5AGTLS
				P10: 
MSExchange TransportService	Worker process with process ID 6908 has exited.
MSExchange TransportService	The service has stopped.
```

**Comment by AlxVD** [comment]
Does anyone has some experience with WinDbg (to check dump file)?

**Comment by Pro** [comment]
You can try the command `!analyze -v`. I already tried it, but the debugging symbols are missing (probably because the analysis has to be run on a Windows 2008 Server...)

**Comment by AlxVD** [comment]
As far as I remember:
1. !analyze won't help with .Net process (and no one has symbols for edgetransport.exe; but you can load symbols for ExchangeDkimSigner.dll).
2. There were some specific extensions for WinDbg that can assist with .Net.

**Comment by rbfajardo** [comment]
The call stack in the crash dump throws an exception. Can't share anything else, don't know what I am looking at either. 

System.Security.Cryptography.CryptographicException.ThrowCryptogaphicException

**Comment by AlxVD** [comment]
Ok. Today it crashed without my help (nobody tried to restart service).
At 0:00:07 and 0:00:08 there were 2 same errors in App-log:
"Couldn't load the key 'C:\Program Files\Exchange DkimSigner\keys\my-dkim-priv.pem' for domain unium.ru. Error message: Invalid DER format for key. (The process cannot access the file because it is being used by another process.
)"
After that at 0:19:39 edgetransport.exe crashed with "CryptographicException". But now I have stack dumped: https://cloud.mail.ru/public/EFu5/tBEzt8ufR

**Comment by Pro** [comment]
Can you upgrade to version 2.1.8
We changed a lot there. This may also fix the last exception you mentioned (see #109)

**Comment by AlxVD** [comment]
No. Still have exception on service stop.

**Comment by Pro** [comment]
@AlxVD does the exception occur every time you stop the service? Or just randomly?

Do you have access to a Visual Studio installation and could try to build the .dll in Debug mode, then start exchange, attach the Visula Studio debugger and get the exact line causing the problem?

**Comment by Pro** [comment]
@AlxVD any updates?
v3.0.0 beta is now released. Maybe you can try this
https://github.com/Pro/dkim-exchange/releases/tag/v3.0.0-beta

**Comment by Pro** [comment]
Fix will be included in version 3.0.0
